### PR TITLE
feat: 16022 Aggregate bls keys and tss shares

### LINKED ISSUE #16022: Aggregate BlsPrivateKeys, BlsPublicKeys and BlsSignatures
* BlsKeys and signatures should allow aggregation based on the addition of the elements.
* Specifically for TSS, we should also allow weighted sum aggregation by calculating the Lagrange coefficient and multiplying it by the elements and their id.

### PR DESCRIPTION
**Pull request by mxtartaglia-sl**
**Description**:
This allows BlsKeys and signatures to be aggregated based on the addition of the elements.
For TSS, Share keys and signatures are aggregated by calculating the weighted sum aggregation by calculating the Lagrange coefficient and multiplying it by the elements and their IDs.

**Related issue(s)**:

Fixes: https://github.com/hashgraph/hedera-services/issues/16022

**Notes for reviewer**:
Many changes on the `TssService` interface.
The future work is related to the following ticket: https://github.com/hashgraph/hedera-services/issues/16096

### CODE REVIEW

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java [hunk 1]
```diff
@@ -35,25 +36,36 @@ public record BlsKeyPair(@NonNull BlsPrivateKey privateKey, @NonNull BlsPublicKe
      * @param publicKey  the public key
      */
     public BlsKeyPair {
-        Objects.requireNonNull(privateKey, "privateKey cannot be null");
-        Objects.requireNonNull(publicKey, "publicKey cannot be null");
+        Objects.requireNonNull(privateKey, "element cannot be null");
+        Objects.requireNonNull(publicKey, "element cannot be null");
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java]: why change the message if the variable names are not changed?

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java [hunk 2]
```diff
@@ -49,11 +65,45 @@ public static BlsPublicKey fromBytes(@NonNull final byte[] bytes) {
...
-     * @throws IllegalArgumentException if the signature representation is invalid
+     * Aggregates multiple {@link BlsPublicKey} into a single {@link BlsPublicKey} for efficient verification.
+     *<p>
+     * This method combines multiple public keys into a single aggregated
+     * public key, which retains the same size as a regular {@link BlsPublicKey} .
+     * The aggregation is performed using elliptic curve point addition in the group defined by each signature schema,
+     * where each signature is a point on the curve.
+     *<p>
+     * An aggregated signature is indistinguishable from a non-aggregated signature in terms of size reducing the
+     * computational cost of verification.
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java]: this mentions signatures, but keys are being aggregated

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java [hunk 3]
```diff
@@ -49,11 +65,45 @@ public static BlsPublicKey fromBytes(@NonNull final byte[] bytes) {
...
+     *
+     * @param publicKeys A list of {@link BlsPublicKey}, where each signature is a point in the elliptic
+     *                   curve group.
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java:77]: again it says signature

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java [hunk 4]
```diff
@@ -49,11 +65,45 @@ public static BlsPublicKey fromBytes(@NonNull final byte[] bytes) {
...
+     * @throws IllegalArgumentException if the publicKeys schemas do not match.
+     */
+    public static BlsPublicKey aggregate(@NonNull final List<BlsPublicKey> publicKeys) {
+        if (Objects.requireNonNull(publicKeys, "publicKeys must not be null").size() < 2) {
+            throw new IllegalArgumentException("Not enough publicKeys to aggregate");
+        }
+        final Collection<SignatureSchema> s =
+                publicKeys.stream().map(BlsPublicKey::signatureSchema).collect(Collectors.toSet());
+        if (s.size() > 1) {
+            throw new IllegalArgumentException("publicKeys must not contain more than one schema");
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java]: I think a better message would be "All keys should have the same schema"

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java [hunk 5]
```diff
@@ -49,11 +65,45 @@ public static BlsPublicKey fromBytes(@NonNull final byte[] bytes) {
...
+     * @throws NullPointerException if signatures is null.
+     * @throws IllegalArgumentException if there are not enough publicKeys to aggregate.
+     * @throws IllegalArgumentException if the publicKeys schemas do not match.
+     */
+    public static BlsPublicKey aggregate(@NonNull final List<BlsPublicKey> publicKeys) {
+        if (Objects.requireNonNull(publicKeys, "publicKeys must not be null").size() < 2) {
+            throw new IllegalArgumentException("Not enough publicKeys to aggregate");
+        }
+        final Collection<SignatureSchema> s =
+                publicKeys.stream().map(BlsPublicKey::signatureSchema).collect(Collectors.toSet());
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java]: I think there is no need to collect, `distinct().count()` should be enough

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsSignature.java [hunk 6]
```diff
@@ -51,31 +67,67 @@ public static BlsSignature fromBytes(@NonNull final byte[] bytes) {
...
+     * computational cost of verification.
+     *
+     * @param signatures A list of {@link BlsSignature}, where each signature is a point in the elliptic
+     *                   curve group.
+     * @return A single aggregated BLS signature.
+     * @throws NullPointerException if signatures is null.
+     * @throws IllegalArgumentException if there are not enough signatures to aggregate.
+     * @throws IllegalArgumentException if the signature schemas do not match.
+     */
+    public static BlsSignature aggregate(@NonNull final List<BlsSignature> signatures) {
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsSignature.java:86]: public key and signature seem to be very similar. do you think its worth deduplicating some code? having methods like aggregate be the same for both?

File: cryptography/hedera-cryptography-blskeygen/src/main/java/com/hedera/cryptography/blskeygen/KeyGen.java [hunk 7]
```diff
@@ -35,11 +35,11 @@
  * <pre>{@code  --help}</pre>
  *
  * <p>Generating a Key Pair:
- * <pre>{@code generate-keys path/to/privateKey.pem path/to/publicKey.pem}</pre>
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-blskeygen/src/main/java/com/hedera/cryptography/blskeygen/KeyGen.java]: the rename in this file seems wrong, looks like intellij did more than it should have

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssPrivateShare.java [hunk 8]
```diff
@@ -36,6 +37,28 @@ public record TssPrivateShare(@NonNull TssShareId shareId, @NonNull BlsPrivateKe
...
+     * Creates a new instance.
+     *
+     * @param id id
+     * @param privateKey the private key
+     * @return a new {@link TssPrivateShare}
+     */
+    public static TssPrivateShare of(final int id, @NonNull final BlsPrivateKey privateKey) {
+        requireNonNull(privateKey, "privateKey must not be null");
+        return new TssPrivateShare(new TssShareId(privateKey.element().field().fromLong(id)), privateKey);
+    }
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssPrivateShare.java]: is there some constraint on the id, like it must be positive?

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssService.java [hunk 9]
```diff
@@ -92,54 +100,6 @@ List<TssPrivateShare> decryptPrivateShares(
      * @throws IllegalStateException if there aren't enough messages to meet the threshold
      */
     @NonNull
-    List<TssPublicShare> computePublicShares(
+    List<TssPublicShare> obtainPublicShares(
             @NonNull TssParticipantDirectory participantDirectory, @NonNull List<TssMessage> validTssMessages);
-
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssService.java:97]: why more these methods?

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java (hunk 1 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java]: My bad, wrong replacement. Thanks for spotting

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java (hunk 2 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java]: c&p. Thanks!

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java (hunk 5 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsPublicKey.java]: Well, i do use the collection after but I can still use the original one

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsSignature.java (hunk 6 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsSignature.java:86]: All of them are super similar. If they implement some generic interface, the three behaviors can fit into only one class. A middle ground is what you said, making the signature and public key generic.
We can create a task for this.

File: cryptography/hedera-cryptography-blskeygen/src/main/java/com/hedera/cryptography/blskeygen/KeyGen.java (hunk 7 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-blskeygen/src/main/java/com/hedera/cryptography/blskeygen/KeyGen.java]: Agree, did not spot this. thanks

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssPrivateShare.java (hunk 8 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssPrivateShare.java]: Yes, and we should enforce it here.

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssService.java (hunk 9 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssService.java:97]: To maintain a consistent approach between the BLS library and the TSS library.
In the Bls library, it is the responsibility of the Key class to aggregate instances of keys.
In TSS, it was the responsibility of the service class. Given that the shares are Bls Keys (or wrappers), it is best to keep a consistent approach.
It seems also better from a client perspective, as they don't need access to an instance of the service class anymore to do this aggregation.

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsSignature.java (hunk 6 above)
Comment (lpetrovic05) [reply cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsSignature.java:86]: created

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java [hunk 10]
```diff
@@ -38,22 +39,33 @@ public record BlsKeyPair(@NonNull BlsPrivateKey privateKey, @NonNull BlsPublicKe
         Objects.requireNonNull(privateKey, "privateKey cannot be null");
         Objects.requireNonNull(publicKey, "publicKey cannot be null");
         if (privateKey.signatureSchema() != publicKey.signatureSchema()) {
-            throw new IllegalArgumentException("The private key and public key must have the same signature schema");
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java:41]: I think this is also a bad rename

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java [hunk 11]
```diff
@@ -38,22 +39,33 @@ public record BlsKeyPair(@NonNull BlsPrivateKey privateKey, @NonNull BlsPublicKe
...
+            throw new IllegalArgumentException("The private key and public key must have the same element schema");
         }
     }
 
     /**
      * Generates a Key Pair (private and public keys)
      *
-     * @param signatureSchema the signature schema to use
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java:48]: looks like bad rename

File: cryptography/hedera-cryptography-bls/src/test/java/com/hedera/cryptography/bls/SignaturesLibraryTest.java [hunk 12]
```diff
@@ -230,7 +230,7 @@ void flipEachBitAndConsume(@NonNull final byte[] original, final @NonNull Consum
                 copy[i] = originalByte;
             }
         }
-        // REVIEW flipping the last bit of the sign of the last element produces the same signature
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/test/java/com/hedera/cryptography/bls/SignaturesLibraryTest.java]: looks like bad rename

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/SignatureSchema.java [hunk 13]
```diff
@@ -88,7 +88,7 @@ public GroupAssignment getGroupAssignment() {
     /**
      * Returns the signature scheme encoded in the byte array
      *
-     * @param bytes the array containing the representation in the first element
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/SignatureSchema.java:91]: looks like bad rename

**Review by lpetrovic05 — APPROVED** [review]

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java (hunk 10 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java:41]: Will Fix in 15211

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java (hunk 11 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/BlsKeyPair.java:48]: Will Fix in 15211

File: cryptography/hedera-cryptography-bls/src/test/java/com/hedera/cryptography/bls/SignaturesLibraryTest.java (hunk 12 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-bls/src/test/java/com/hedera/cryptography/bls/SignaturesLibraryTest.java]: Will Fix in 15211

File: cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/SignatureSchema.java (hunk 13 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-bls/src/main/java/com/hedera/cryptography/bls/SignatureSchema.java:91]: Will Fix in 15211

**Pull request merged**